Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,170 @@
+		} catch let error {
+			let carthageError: CarthageError = (error as? CarthageError) ?? CarthageError.internalError(description: error.localizedDescription)
+			result = .failure(carthageError)
+		}
+		return SignalProducer(result: result)
+	}
+
+	private func traverse(dependencies: [(Dependency, VersionSpecifier)]) throws {
+		for (dependency, versionSpecifier) in dependencies {
+			let versionSet = try findAllVersions(for: dependency, compatibleWith: versionSpecifier)
+			for version in versionSet {
+				let pinnedDependency = PinnedDependency(dependency: dependency, pinnedVersion: version)
+
+				if !handledDependencies.contains(pinnedDependency) {
+					handledDependencies.insert(pinnedDependency)
+
+					let transitiveDependencies = try findDependencies(for: dependency, version: version)
+					try traverse(dependencies: transitiveDependencies)
+				}
+			}
+		}
+	}
+
+	private func findAllVersions(for dependency: Dependency, compatibleWith versionSpecifier: VersionSpecifier) throws -> [PinnedVersion] {
+		do {
+			let versionSet: [PinnedVersion]
+			if let cachedVersionSet = versionSets[dependency] {
+				versionSet = cachedVersionSet
+			} else {
+				var pinnedVersions = [PinnedVersion]()
+				let pinnedVersionsProducer: SignalProducer<PinnedVersion, CarthageError>
+				var gitReference: String?
+
+				switch versionSpecifier {
+				case .gitReference(let hash):
+					pinnedVersionsProducer = resolvedGitReference(dependency, hash)
+					gitReference = hash
+				default:
+					pinnedVersionsProducer = versionsForDependency(dependency)
+				}
+
+				let concreteVersionsProducer = pinnedVersionsProducer.filterMap { pinnedVersion -> PinnedVersion? in
+					pinnedVersions.append(pinnedVersion)
+					return nil
+				}
+
+				_ = try concreteVersionsProducer.collect().first()!.dematerialize()
+
+				versionSets[dependency] = pinnedVersions
+
+				let storedDependency = self.dependencyMappings?[dependency] ?? dependency
+				try localRepository?.storePinnedVersions(pinnedVersions, for: storedDependency, gitReference: gitReference)
+
+				versionSet = pinnedVersions
+			}
+
+			let filteredVersionSet = versionSet.filter { pinnedVersion -> Bool in
+				versionSpecifier.isSatisfied(by: pinnedVersion)
+			}
+
+			delegate?.diagnosticResolver(self, didFindVersions: filteredVersionSet, compatibleWith: versionSpecifier, for: dependency)
+
+			return filteredVersionSet
+		} catch let error {
+			delegate?.diagnosticResolver(self, didEncounterError: error, whileRetrievingVersionsFor: dependency, compatibleWith: versionSpecifier)
+
+			if ignoreErrors {
+				return [PinnedVersion]()
+			} else {
+				throw error
+			}
+		}
+	}
+
+	private func findDependencies(for dependency: Dependency, version: PinnedVersion) throws -> [(Dependency, VersionSpecifier)] {
+		do {
+			let transitiveDependencies: [(Dependency, VersionSpecifier)] = try dependenciesForDependency(dependency, version).collect().first()!.dematerialize()
```

Comment:
This would crash when the dependency has no dependency.

Reply:
I am not quite sure it is such a great idea to let this reactive style of programming trickle through to every component (rather than in the assembly part of the app, glueing the different components together). Not only does every developer need to understand the reactive patterns instead of just ordinary Swift, it also makes the API more cumbersome in my opinion. 

The above may be due to me having too little experience with this idiom. However, as can be witnessed by the resolver performance issues, it is hard to get a grasp on algorithm time complexities while using reactive patterns everywhere.

That being said, I'm open to concrete suggestions how to improve the code in a reactive manner if that really is the wish. Maybe you could propose how to rewrite the code above?

Reply:
I am not so religious about the reactive style but that has been the standard so far. This is a question for @mdiep .